modules/actions/pool/v3/upsert-pools.ts

```diff
@@ -44,16 +45,46 @@ export const upsertPools = async (
         console.error('Error creating tokens', e);
     }
 
-    // Get the data for the tables about pools
-    const dbPools = subgraphPools
-        .map((poolData) => subgraphPoolUpsert(poolData, onchainData[poolData.id], chain, blockNumber))
-        .filter((item): item is Exclude<SubgraphPoolUpsertData, null> => Boolean(item));
+    // Get the prices
+    const prices = await prisma.prismaTokenCurrentPrice
+        .findMany({
+            where: {
+                chain: chain,
+                tokenAddress: { in: allTokens.map((token) => token.address) },
+            },
+        })
+        .then((prices) => Object.fromEntries(prices.map((price) => [price.tokenAddress, price.price])));
 
-    // Enrich updates with USD values
-    const poolsWithUSD = await poolUpsertsUsd(dbPools, chain, allTokens);
+    const pools = subgraphPools
+        .map((fragment) => poolUpsertTransformerV3(fragment, chain, blockNumber))
+        .map((upsert) => {
+            const update = applyOnchainDataUpdateV3(
+                onchainData[upsert.pool.id],
+                upsert.tokens,
+                chain,
+                upsert.pool.id,
+                blockNumber,
+            );
+            return {
+                ...upsert,
+                poolDynamicData: update.poolDynamicData,
+                poolTokenDynamicData: update.poolTokenDynamicData,
+            };
+        })
+        .map((upsert) => {
+            const update = enrichPoolUpsertsUsd(
+                { poolDynamicData: upsert.poolDynamicData, poolTokenDynamicData: upsert.poolTokenDynamicData },
+                prices,
+            );
+            return {
+                ...upsert,
+                poolDynamicData: update.poolDynamicData,
+                poolTokenDynamicData: update.poolTokenDynamicData,
+            };
+        });
```

Comment:
```suggestion
        .map((upsert) => enrichPoolUpsertsUsd(
                { poolDynamicData: upsert.poolDynamicData, poolTokenDynamicData: upsert.poolTokenDynamicData },
                prices,
            ));
```

Reply:
same, need to add the original immutable part to the return value